### PR TITLE
bug(core): Revert _step_ predefined key since we already use it; fix binary joins

### PR DESCRIPTION
**Pull request by vishramachandran**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

**Current behavior :** (link exiting issues here : https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests)

`_step_` tag was earlier introduced as predefined tag. As a result new partKeys are created which aren't equivalent to old ones. In addition, binary joins have issues since the one-side in OneToOne or OneToMany joins now have duplicate join keys.

**New behavior :**

* Rename `_step_` tag with unused `_pi_` (stands for publish interval). Test case included to ensure name update works fine.
* Include tags like `_step_` and `_pi_` in the `on` specifier of joins

### CODE REVIEW

File: query/src/main/scala/filodb/query/exec/BinaryJoinExec.scala [hunk 1]
```diff
@@ -51,11 +51,11 @@ final case class BinaryJoinExec(queryContext: QueryContext,
   require(!on.contains(metricColumn), "On cannot contain metric name")
 
   val onLabels = on.map(Utf8Str(_)).toSet
+  val withExtraOnLabels = onLabels ++ Seq("_pi_", "_step_") // needed since these are internal tags
```
Comment (tjackpaul) [inline query/src/main/scala/filodb/query/exec/BinaryJoinExec.scala:54]: Adding only new tag ‘pi’ is not sufficient? 

Also, do we have to keep this forever? Or for a short term would suffice?

**Review by tjackpaul — APPROVED** [review]
lgtm.. thanks for the quick fix.. 

had a question.. added the same in comment

File: core/src/main/resources/filodb-defaults.conf [hunk 2]
```diff
@@ -25,9 +25,12 @@ filodb {
...
     # Predefined keys allow space to be saved for over the wire tags with the given keys
-    # WARN: It is suggested to only ADD to predefined keys.
-    #       Changing/Renaming existing keys will cause incorrect tags to be reported and probably require a clean wipe.
-    predefined-keys = ["_ws_", "_ns_", "app", "__name__", "instance", "dc", "le", "job", "exporter", "_step_"]
+    # WARN: 1. It is suggested to only ADD to predefined keys that are not used already,
+    #          which means only new tags with underscore prefixed and suffixed.
+    #       2. Changing/Renaming existing keys will cause incorrect tags to be reported and probably require a clean wipe.
+    #       3. Even when adding new keys, be careful. If that tag is already being used, there is no
+    #          equality of partKeys between old and new time series.
+    predefined-keys = ["_ws_", "_ns_", "app", "__name__", "instance", "dc", "le", "job", "exporter", "_pi_"]
```
Comment (broneill) [inline core/src/main/resources/filodb-defaults.conf:33]: Comments for what these special keys mean? I cannot find any references to `_pi_`

File: query/src/main/scala/filodb/query/exec/BinaryJoinExec.scala (hunk 1 above)
Comment (vishramachandran) [reply query/src/main/scala/filodb/query/exec/BinaryJoinExec.scala:54]: Technically yes. But this issue will also occur when customers change the publish interval themselves. Joins will fail and they have to provide this internal tag explicitly. Instead, we could do it automatically for them.

File: core/src/main/resources/filodb-defaults.conf (hunk 2 above)
Comment (vishramachandran) [reply core/src/main/resources/filodb-defaults.conf:33]: `_pi_` stands for Publish interval. It is intentionally a new unused tag name so the new time series can be differentiated from the ones that already have `_step_` tag.

Comment (vishramachandran) [reply core/src/main/resources/filodb-defaults.conf:33]: Will add docs.

**Comment by velvia** [comment]
I strongly recommend that predefined keys NOT be removed from the list in defaults.   Change the gateway instead to not use it, but removing something creates the possibility that there could exist two possible meanings in persisted part keys, which could lead to corruption.   While you could use deployment controls etc to guard against this, as a matter of principle, the predefined-keys list should be treated as append-only.

**Review by TanviBhavsar — APPROVED** [review]

**Comment by vishramachandran** [comment]
@velvia Thanks for the heads-up. We were aware of the consequences while proceeding with this. It was a necessary change.

**Pull request merged**